xdsl/dialects/onnx.py

```diff
@@ -256,6 +256,98 @@ def verify_(self) -> None:
             )
 
 
+@irdl_op_definition
+class Reshape(IRDLOperation):
+    """
+    Reshape the input tensor similar to numpy.reshape.
+    First input is the data tensor, second input is a shape tensor which specifies the output shape. It outputs the reshaped tensor.
+    At most one dimension of the new shape can be -1. In this case, the value is
+    inferred from the size of the tensor and the remaining dimensions. A dimension
+    could also be 0, in which case the actual dimension value is unchanged (i.e. taken
+    from the input tensor). Shape (second input) could be an empty shape, which means converting to a scalar.
+    The input tensor's shape and the output tensor's shape are required to have the same number of elements.
+    """
+
+    name = "onnx.Reshape"
+    T = Annotated[AnyFloat | IntegerType, ConstraintVar("T")]
+    data = operand_def(TensorType[T])
+    shape = operand_def(TensorType)
```

Comment:
The shape in onnx is a tensor of 64-bit signless integer values, so I think constraining it at least to an integer here is better than leaving it as a tensor with arbitrary element types.

Reply:
This is checked in `verify_` as @kayode-gif has implemented ATM. We are just missing a test case in invalid ops for lit.

Reply:
Ok. Based on previous reviews, it's still unclear (to me) when the actual definition should be annotated with types and when most checking should be deferred to `verify_`. I would just like to understand better why it's preferable not to use a type constraint for the shape here (and instead check in `verify_`).

Reply:
My understanding is that the bitwidth is not part of the type in Python "typeland" and in the previous formulation in this PR it was just a Python instance; Hence cannot be checked as part of the type during type inference.
I vaguely remember having trouble with this when I needed some similar restriction on types in the past.

@superlopuh or @math-fehr might be able to explain this better.

Reply:
I agree, with @brnorris03, constraining it to an IntegerAttr seems like a good idea, and that should be supplemented with a check for 64-bit signless integer in the `verify_` method